### ISSUE TITLE
Joining different labels

### ISSUE THREAD

**Issue by fabxc** [prometheus/prometheus#2204]
If we are monitoring complex systems, we might get metrics on different connected entities from different metric providers. The label naming might not be 100% aligned, i.e. in Kubernetes we get container metrics with a `pod_name` label from cAdvisor and metrics with a `pod` label from kube-state-metrics.
The values are the same pod identifying names. However, we have to use `label_replace()` to be able to join them properly, which is always somewhat awkward and hard to decipher.

I think this will be a common case across many systems and first-class support to join such labels could be beneficial. I imagine a syntax along the lines of:

```
container_last_seen and on(pod_name=pod) kube_pod_info
```

Where the names left and right of `=` refer to the matched on labels of the left and right metrics respectively.
I'd hope it to be easy enough to implement in our parser and think it's generally an rather obvious syntax for the semantics.

I know that's not a trivial extension to PromQL, so this is just to start a discussion.

@brian-brazil

**Comment by brian-brazil** [comment]
The correct answer here is to fix the exporter, I consider label_replace and what you propose here only as stop-gaps until that can happen.

**Comment by fabxc** [comment]
I figured that would be the go-to response.

I agree with that when it comes to labels containing concatenated information we have to split before joining etc. – but it's just an unrealistic expectation to align 3+ components written by different teams. Especially if those have different stability concerns and won't simply accept breaking their metrics used by thousands of people.
Neither is it to have all those users use `label_replace` in dozens of alerts and especially not in ad-hoc queries.

**Comment by brian-brazil** [comment]
There's also the 3rd option of metric_relabel_configs.

I'm used to it being really uncommon for you to want to join data across jobs, let alone with different labels.

**Comment by fabxc** [comment]
I see it to be really common actually – e.g. cAdvisor exports container metrics labeled with pod name. Kubelet exports pod info metrics (at least I think that's the best source, issue pending on that), kube-state-metrics exporting metrics on abstract objects like deployments, replica sets, etc.
I'd generally take all those together and assemble some recording rules providing me with comprehensive metrics on resource usage by a deployment, through pods belonging to them, and containers belonging to those.

These would mostly be recording rules, yes – they'd be fairly static and label_replace would be okay. However, in the end if I want to drill down I'd have to use that ad-hoc again and again.
I'm not a fan of `metric_relabel_configs` as they alter the state of well-documented metrics. Prometheus doesn't have the usability facilities to make up for that, i.e. proper query-time completion on potential labels per metric and their values, ...

**Comment by brian-brazil** [comment]
In the systems I'm used to, the equivalent pod data and instance data would almost never be joined - it wasn't needed in practice to actually link things together end to end to do the alerting you wanted. k8 is a more complicated system though.

This does sound like something sig-instrumentation should be standardising.

> Would this addition have any weird edge-case semantics or be confusing?

I've only been thinking on it a few minutes, but I see potential issues around e.g. `without` - how do you spell it then? What are the semantics?

Which of the two labels does the output metric have? Do we break the commutivity we currently have? How does this interact with group_left?

Label matching is already confusing to users, so I'm wary of adding more modifiers.

The only place I've ever seen anything like this is for alertmanager inhibits, for when you want to use an alert another team has setup for this purpose but you called your zone/cluster/datacenter label something else.

A 4th option is to add `label_copy`, as `label_replace` is unwieldy for that simple case.

**Comment by fabxc** [comment]
Yes, I'm trying to think about standardisation of k8s metrics but see exactly those differing requirements of metric stability.

As it would be an addition, I don't think we'd break anyone. This would only affect `on()`. 
We'd indeed have to pick one side, probably left, so it would break commutativity, which isn't nice. I agree that matching is already confusing to many and added complexity is not great.

`label_copy` might be a step that's easier to understand for sure.

We might want to add that and see over a few weeks how far standardisation efforts get us.

**Comment by brian-brazil** [comment]
> This would only affect on(). 

We'd need it with `ignoring` too, as in principle you should be favouring `ignoring` over `on` to produce generic shareable rules. The alternative would involve this descending into a 2nd form of group_left where we keep all the labels from one side.

label_copy is in #959

**Comment by beorn7** [comment]
On 18 November 2016 at 12:57, Fabian Reinartz <notifications@github.com>
wrote:

> However, in the end if I want to drill down I'd have to use that ad-hoc
> again and again.
> [...] Prometheus doesn't have the usability facilities to make up for
> that, i.e. proper query-time completion on potential labels per metric and
> their values, ...
>
As a side note: I think we should really improve on the auto completion in
the expression browser. We were quite far already, and essentially gave up
because of a trivial usability quirk. If I hadn't so many other things on
my plate, I'd work on it, despite having no real expertise in UIs.

And if proper auto completion would solve the issue discussed here, I'd
prefer it over making PromQL more complex.

-- 
Björn Rabenstein, Engineer
http://soundcloud.com/brabenstein

SoundCloud Ltd. | Rheinsberger Str. 76/77, 10115 Berlin, Germany
Managing Director: Alexander Ljung | Incorporated in England & Wales with
Company No. 6343600 | Local Branch Office | AG Charlottenburg  | HRB 110657B

**Comment by brian-brazil** [comment]
Auto-completion is a good thing, but I don't see it helping much with the fundamental issue here.

**Comment by fabxc** [comment]
Let's say it would alleviate some of the pain of rarely having a clue what labels a given metric has and not being able to see its help text. and hence relying more strongly on application's documentation of its metrics. That return might make people more reluctant to break their metrics.

OTOH, this only helps Prometheus, which is not the only system applications export metrics for.

**Comment by davidquarles** [comment]
@fabxc @brian-brazil among the options listed, which do you recommend?  below is what i'm currently doing to enable the exact workflow [@fabxc mentioned](https://github.com/prometheus/prometheus/issues/2204#issuecomment-261515520), which in our case supports dashboards, alerts, and ad-hoc queries.  I'm curious what the footprint of and tradeoffs between the various options are.
```yaml
metric_relabel_configs:

# supports joins to node-exporter metrics
- source_labels: [kubernetes_name, node]
  regex: 'kube-state-metrics;(.+)'
  target_label: instance

# supports joins to cadvisor metrics
- source_labels: [kubernetes_name, pod]
  regex: 'kube-state-metrics;(.+)'
  target_label: pod_name

# supports label-based, workload-level queries
- action: labelmap
  regex: label(_k8s|)_app
  replacement: app

# supports owner-based, workload-level queries
- source_labels: [kubernetes_name, __name__, owner_kind, owner_name]
  regex: 'kube-state-metrics;kube_pod_owner;(.+);(.+)'
  replacement: '$1/$2'
  target_label: owner

# supports owner-based queries targeting the parent deployment
- source_labels: [kubernetes_name, __name__, owner_kind, owner_name]
  regex: 'kube-state-metrics;kube_pod_owner;ReplicaSet;(.+)-(.+)'
  replacement: 'Deployment/$1'
  target_label: owner
```
```yaml
groups:
- name: pod-info
  rules:
  - record: kube_pod_info_ex
    expr: |
      # fall back to app label for static/orphaned pods
      label_replace(
        kube_pod_labels
        * on(namespace, pod) group_left(host_ip, node, pod_ip)
        kube_pod_info
        * on(namespace, pod) group_left (owner_is_controller, owner, owner_kind, owner_name)
        kube_pod_owner{owner_is_controller="<none>"},
        "owner", "$0", "app", ".*"
      )
      or on (namespace, pod)(
        kube_pod_labels
        * on(namespace, pod) group_left(host_ip, node, pod_ip)
        kube_pod_info
        * on(namespace, pod) group_left (owner_is_controller, owner, owner_kind, owner_name)
        kube_pod_owner{owner_is_controller="true"},
      )
- name: pod-aggregates
  rules:
  - record: pod_memory_usage_bytes
    expr: |
      kube_pod_info_ex
      * on (namespace, pod_name) group_left
      sum by (namespace, pod_name)(container_memory_usage_bytes)
  - record: pod_spec_memory_limit_bytes
    expr: |
      kube_pod_info_ex
      * on (namespace, pod_name) group_left
      sum by (namespace, pod_name)(container_spec_memory_limit_bytes)
- record: pod_spec_memory_request_bytes
  expr: |
    kube_pod_info_ex
    * on (namespace, pod_name) group_left
    sum by (namespace, pod_name)(kube_pod_container_resource_requests_memory_bytes)
    or on (namespace, pod_name)
    kube_pod_info_ex > bool 1
  # other resources...
- name: workload-aggregates
  # ...
- name: node-aggregates
  # ...
```

**Comment by brian-brazil** [comment]
Fixing the source data is the best option, but failing that metric_relabel_configs is best.

**Comment by chancez** [comment]
I'd be super happy if a `label_copy` was added, that's all that's really needed here, and potentially a `label_remove`, though that's not as necessary for the case of joining.

**Comment by tangxuye** [comment]
I wonder if some scenes that we need add some labels to the metrics of the third-exporter.For example, the node_exporter,if I want to add one or more labels to the metrics.What should I do?

**Comment by gjcarneiro** [comment]
I have the same problem, I need to join `kube_pod_labels`, which has a `pod` label containing the pod name, with the cAdvisor metrics such as `container_cpu_usage_seconds_total`, which has a `pod_name` label, same value but different label name.

**Comment by Sennahoi** [comment]
I would love the requested feature as well.

**Comment by FakeEmperor** [comment]
I would also like to see this (being able to do `on()` and `ignoring()` with matching different labels) implemented.

**Comment by ahothan** [comment]
+1

**Comment by mpfgomes** [comment]
+1

**Comment by geekofalltrades** [comment]
I would like to be able to join metrics from node-exporter with node metrics from kube_state_metrics.

node-exporter only provides the node name, which in EC2 is something like `ip-10-224-40-127.us-west-2.compute.internal` -- not human-readable. The `kube_node_labels` metric has all of the juicy labels that we apply, many of which _are_ human-readable and would be better identifiers.

These two metrics align on that node name, where node-exporter metrics have a `kubernetes_pod_node_name` label that matches the `node` label of the `kube_node_labels` metric from kube_state_metrics. It would be great to be able to join on these differently-named labels so we can smash more identifiable information from kube_state_metrics back into the node-exporter metrics.

**Comment by roidelapluie** [comment]
`kube_metric1 and on (kubernetes_pod_node_name) label_join(kube_node_labels, "kubernetes_pod_node_name","","node")` is basically the `label_copy` function suggested in https://github.com/prometheus/prometheus/issues/2204#issuecomment-407248342

It was merged on Jun 16, 2017, so after this ticket was created, and resolves this ticket.

**Comment by brian-brazil** [comment]
True, that'd also cover it (though not quite what that function is for). Kubernetes is also fixing their metrics, but is awaiting their depreciation cycle. Fixing the labels before they get to vector matching is a better option than making vector matching even more complicated in any case.

**Comment by RichiH** [comment]
As per our dev summit consensus, we are re-opening this issue.

**Comment by roidelapluie** [comment]
Here is a design doc / proposal

https://docs.google.com/document/d/1KWh7TNt66zMqn6_yBytZ_n_KEXOhHX0Qmmv2JLssy0s/edit?usp=sharing

**Comment by danthegoodman1** [comment]
Any update on this? I want to divide 2 metrics where one is `exported_pod`, and another uses `name` but am unable to.

**Comment by RichiH** [comment]
The status we have is in the design doc. Anyone willing to help with a PR, please contact us.

**Comment by sirkubax** [comment]
any chance we could have a regex on the on() join mathing?



my example
```
kube_endpoint_info{endpoint="accounts",instance="k8s-infra-1.test.kv",job="kubernetes-service-endpoints",k8s_app="kube-state-metrics",kubernetes_namespace="kube-system",kubernetes_node="k8s-infra-1.test.kv",namespace="cards-dev"}  1
```
combine with:
```
elasticsearch_indices_store_size_bytes_total{app="elasticsearch-exporter",chart="elasticsearch-exporter-3.0.0",cluster="cluster_1",heritage="Helm",index="k8s-namespace-cards-dev-2021.04.16",instance="k8s-infra-1.test.kv",job="kubernetes-service-endpoints",kubernetes_namespace="monitoring",kubernetes_node="k8s-infra-1.test.kv"}  1989206426
```

The only aligning match I could have in second query is 

index="k8s-namespace-**cards-dev**-2021.04.16"

and I want to match it with
namespace="cards-dev" from the first query


so I'd expect sth like a label_replace or label_copy
or 
(pseudocode):

on(namespace == ($1 of index{k8s-namespace-(.*)-[0-9-]+} )

**Comment by sirkubax** [comment]
https://devops.stackexchange.com/questions/10941/prometheus-join-series-based-on-the-key-part-of-a-label
related to partial label match

**Comment by roidelapluie** [comment]
Then label_replace seems your best option.

**Comment by lebenitza** [comment]
> I would like to be able to join metrics from node-exporter with node metrics from kube_state_metrics.
> 
> node-exporter only provides the node name, which in EC2 is something like `ip-10-224-40-127.us-west-2.compute.internal` -- not human-readable. The `kube_node_labels` metric has all of the juicy labels that we apply, many of which _are_ human-readable and would be better identifiers.
> 
> These two metrics align on that node name, where node-exporter metrics have a `kubernetes_pod_node_name` label that matches the `node` label of the `kube_node_labels` metric from kube_state_metrics. It would be great to be able to join on these differently-named labels so we can smash more identifiable information from kube_state_metrics back into the node-exporter metrics.

I want to revive the discussion for this ^ exact use case. `label_copy` would help a great deal. 

I could not see any hints regarding the status in the design doc or anywhere else.

**Comment by beorn7** [comment]
Note: The [design doc linked above](https://docs.google.com/document/d/1KWh7TNt66zMqn6_yBytZ_n_KEXOhHX0Qmmv2JLssy0s/edit?usp=sharing) is not in https://github.com/prometheus/proposals (not even as one of the issues that tracks converting old design docs to the repo). That should probably happen before anyone seriously tackles this.

Also note that there is work currently ongoing to better treat OTel resource attributes, and/or a special treatment of info metrics, which might take some of the pain away from the use cases described here.

**Comment by eero-t** [comment]
> Also note that there is work currently ongoing to better treat OTel resource attributes, and/or a special treatment of info metrics, which might take some of the pain away from the use cases described here.

@beorn7 Could you provide some links to discussion / design on that?  Use of resource attributes / `target_info` could save a lot of space, but needing to do join and list in those joins _all_ the labels in the queries is rather bothersome, so I'm very much interested on this...

**Comment by beorn7** [comment]
https://docs.google.com/document/d/1ebhGNLs3uhdeprJCullM-ywA9iMRDg_mmnuFAQCloqY/edit?usp=sharing

**Comment by ringerc** [comment]
@eero-t also https://github.com/prometheus/prometheus/pull/15233

**Comment by sylr** [comment]
I asked claude to give it a try: https://github.com/prometheus/prometheus/compare/main...sylr:prometheus:feat/join-different-labels

**Comment by machine424** [comment]
Hello from Bug Scrub!

We think this issue is still relevant, @roidelapluie would you have time to continue working on https://docs.google.com/document/d/1KWh7TNt66zMqn6_yBytZ_n_KEXOhHX0Qmmv2JLssy0s/edit?usp=sharing
